service/matching/taskReader.go

```diff
@@ -127,9 +127,18 @@ func newTaskReader(tlMgr *taskListManagerImpl, isolationGroups []string) *taskRe
 func (tr *taskReader) Start() {
 	tr.Signal()
 	for g := range tr.taskBuffers {
-		go tr.dispatchBufferedTasks(g)
+		g := g
+		tr.stopWg.Add(1)
+		go func() {
+			defer tr.stopWg.Done()
+			tr.dispatchBufferedTasks(g)
+		}()
```

Comment:
there are other ways to achieve this, but tbh I'm fond of "always make concurrency controls visible where it is made concurrent".

alternatively, a `backgroundDispatch(g)` that does this internally (and you do not `go backgroundDispatch(g)`) works pretty well.  that also makes you relatively immune to mutating range vars.